library/ssl_tls.c

```diff
@@ -5957,36 +5957,53 @@ void mbedtls_ssl_conf_curves( mbedtls_ssl_config *conf,
 {
     conf->curve_list = curve_list;
 }
-#endif
+#endif /* MBEDTLS_ECP_C */
 
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
 int mbedtls_ssl_set_hostname( mbedtls_ssl_context *ssl, const char *hostname )
 {
-    size_t hostname_len;
+    /* Initialize to suppress unnecessary compiler warning */
+    size_t hostname_len = 0; 
 
-    if( hostname == NULL )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    /* Check if new hostname is valid before making any change to current one */
 
-    hostname_len = strlen( hostname );
+    if( hostname != NULL )
+    {
+	hostname_len = strlen( hostname );
 
-    if( hostname_len + 1 == 0 )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+	if( hostname_len > MBEDTLS_SSL_MAX_HOST_NAME_LEN )
+	    return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
 
-    if( hostname_len > MBEDTLS_SSL_MAX_HOST_NAME_LEN )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    /* Now it's clear that we will overwrite the old hostname, so free it safely */
 
-    ssl->hostname = mbedtls_calloc( 1, hostname_len + 1 );
+    if( ssl->hostname != NULL )
+    {
+        mbedtls_zeroize( ssl->hostname, strlen( ssl->hostname ) );
+        mbedtls_free( ssl->hostname );
+    }
 
-    if( ssl->hostname == NULL )
-        return( MBEDTLS_ERR_SSL_ALLOC_FAILED );
+    /* Passing NULL as hostname shall clear the old one */
+    
+    if( hostname == NULL )
+    {
+	ssl->hostname = NULL;
+    }
+    else
+    {
+	ssl->hostname = mbedtls_calloc( 1, hostname_len + 1 );
+
+	if( ssl->hostname == NULL )
+	    return( MBEDTLS_ERR_SSL_ALLOC_FAILED );
 
-    memcpy( ssl->hostname, hostname, hostname_len );
+	memcpy( ssl->hostname, hostname, hostname_len );
 
-    ssl->hostname[hostname_len] = '\0';
+	ssl->hostname[hostname_len] = '\0';
```

Comment:
Just an observation, this is not necessary because calloc already zeroizes data. Still, no harm in doing it.